### PR TITLE
feat: prioritized mempool

### PR DESCRIPTION
**Pull request by tzdybal**
This PR brings prioritized mempool from Tendermint v0.34.20 to Optimint.

mempool was copied from tendermint v0.34.20
It was adapted to make it compatible:
 - tmsync package was removed (sync used instead)
 - clist copied to mempool/, imports fixed
 - v0 removed
 - v1/reactor.go removed

Changes in node, rpc, state packages was caused by changes in mempool
API.

Resolves: #191

### CODE REVIEW

File: mempool/ids_test.go [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+// 	require.NoError(t, err)
+
+// 	ids.ReserveForPeer(peerID)
+// 	require.EqualValues(t, 1, ids.GetForPeer(peerID))
+// 	ids.Reclaim(peerID)
+
+// 	ids.ReserveForPeer(peerID)
+// 	require.EqualValues(t, 2, ids.GetForPeer(peerID))
+// 	ids.Reclaim(peerID)
+// }
```
Comment (liamsi) [inline mempool/ids_test.go:23]: Why is this commented out?

File: mempool/ids.go [hunk 2]
```diff
@@ -0,0 +1,3 @@
+package mempool
+
+// These functions were moved into v0/reactor.go and v1/reactor.go
```
Comment (liamsi) [inline mempool/ids.go:3]: What is the purpose of this?

File: mempool/mempool.go [hunk 3]
```diff
@@ -1,43 +1,67 @@
...
 	"fmt"
+	"math"
 
 	abci "github.com/tendermint/tendermint/abci/types"
-	"github.com/tendermint/tendermint/p2p"
 	"github.com/tendermint/tendermint/types"
 )
 
+const (
+	MempoolChannel = byte(0x30)
```
Comment (liamsi) [inline mempool/mempool.go:14]: Isn't this an implementation detail of the tendermint p2p layer?

File: mempool/ids.go (hunk 2 above)
Comment (tzdybal) [reply mempool/ids.go:3]: Copy paste from tendermint. I'll remove those files.

File: mempool/mempool.go (hunk 3 above)
Comment (tzdybal) [reply mempool/mempool.go:14]: Good catch!

File: mempool/ids_test.go (hunk 1 above)
Comment (tzdybal) [reply mempool/ids_test.go]: Copy paste from Tendermint ;)

**Review by liamsi — APPROVED** [review]
utACK

**Pull request merged**